basics/uri.adoc

```diff
@@ -25,253 +26,488 @@ specific language governing permissions and limitations
-
-The standard URI format is defined in https://datatracker.ietf.org/doc/html/rfc3986[RFC3986] and illustrated in the figure below.
+uProtocol uses https://datatracker.ietf.org/doc/html/rfc3986[Uniform Resource Identifiers] (URI) to uniquely identify (and/or address) resources such as devices, software components, topics, methods, etc on a computer network. https://datatracker.ietf.org/doc/html/rfc3986[RFC3986] defines the structure of a URI as follows:
 
 .URI
 image::rfc3986.png[RFC3986 URI]
 
-URI specifications are covered at length in RFC3986 and will not be covered in this document, we will focus on the additions to RFC3986 that are specific to the uProtocol.
+Using this terminology, uProtocol URIs (UUri) can be structured like this:
 
-The following is taken from https://datatracker.ietf.org/doc/html/rfc3986#appendix-A[RFC3986 Appendix-A] and shall be used as reference in this specification:
+.UUri Specification
+[#uuri-specification]
+image::uri.drawio.svg[uProtocol Uri Specification]
 
-[source]
+A UUri can be represented as
+
+* an instantiation of the <<_data_model,UUri Data Model>> in a supported programming language,
+* a URI containing (logical) names only as defined by <<_long_form>>,
+* a URI containing numeric identifiers as defined by <<_short_form>> or
+* a byte array as defined by <<_micro_form>>.
+
+In the remainder of this document, the term _UUri_ is used to refer to an instantiation of the data model, whereas the terms _long form URI_, _short form URI_ and _micro form URI_ are used to refer to its other representations respectively.
+
+== Data Model
+
+The UUri data model is defined in link:../up-core-api/uprotocol/uri.proto[uri.proto] using the proto3 type system.
+The diagram below provides an overview of the types and their relations using UML2 notation.
+
+.UUri Data Model
+[#uuri-data-model]
+[plantuml, format="png", width="100%", id="UUri Object Model"]
 ----
- gen-delims = ":" / "/" / "?" / "#" / "\[" / "\]" / "@"
- sub-delims = "!" / "$" / "&" / "'" / "(" / ")"
- / "*" / "+" / "," / ";" / "="
- pchar = unreserved / pct-encoded / sub-delims / ":" / "@"
- unreserved = ALPHA / DIGIT / "-" / "." / "_" / "~"
- reserved = gen-delims / sub-delims
+@startuml
+class UAuthority {
+  name: string [0..1]
+}
+class UEntity {
+  name: string [0..1]
+  id: uint32 [0..1]
+  version_major: uint32 [0..1]
+  version_minor: uint32 [0..1]
+}
+class UResource {
+  name: string [0..1]
+  instance: string [0..1]
+  message: string [0..1]
+  id: uint32 [0..1]
+}
+class UUri
+
+hide empty members
+
+UAuthority "authority" <--* UUri
+UEntity "entity" <--* UUri
+UResource "resource" <--* UUri
+@enduml
 ----
 
-<<uuri-specification>> diagram below provides an overview of the specifications that we will elaborate in this document. 
+NOTE: All properties of UAuthority, UEntity and UResource are _optional_, whereas all of UUri's aggregation relations have cardinality 1, i.e. the `authority`, `entity` and `resource` properties are _never_ empty.
```

Comment:
In the proto3 file, all of UUri's properties are optional, i.e. they can be empty. In the class model, UUri's properties (represented by its aggregations) are **not** optional because they have cardinality 1